core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -682,14 +695,224 @@ decodeUtf8Lax = decodeUtf8
 -- Decoding Utf16
 -------------------------------------------------------------------------------
 
+{-# INLINE_NORMAL sanitizeW8ToW16Chunks #-}
+sanitizeW8ToW16Chunks
+    :: Monad m
+    => Stream m (Array Word8) -> Stream m (Maybe (Word8, Word8), Array Word8)
+sanitizeW8ToW16Chunks (D.Stream step state) = D.Stream step1 (Nothing, state)
+
+    where
+
+    {-# INLINE_LATE step1 #-}
+    step1 gst (Nothing, st) = do
+        r <- step (adaptState gst) st
+        return $
+            case r of
+                Yield arr st1 ->
+                    let len = Array.length arr
+                     in if (len .&. 1) == 1
+                        then let arr1 = Array.getSliceUnsafe 0 (len - 1) arr
+                                 remElem = Array.getIndexUnsafe (len - 1) arr
+                              in Yield (Nothing, arr1) (Just remElem, st1)
+                        else Yield (Nothing, arr) (Nothing, st1)
+                Skip s -> Skip (Nothing, s)
+                Stop -> Stop
+    step1 gst (Just remElem, st) = do
+        r <- step (adaptState gst) st
+        return $
+            case r of
+                Yield arr st1 ->
+                    let len = Array.length arr
+                     in if (len .&. 1) == 1
+                        then let arr1 = Array.getSliceUnsafe 1 (len - 1) arr
+                                 fstElem = Array.getIndexUnsafe 0 arr
+                                 w16 = (remElem, fstElem)
+                              in Yield (Just w16, arr1) (Nothing, st1)
+                        else let arr1 = Array.getSliceUnsafe 1 (len - 2) arr
+                                 fstElem = Array.getIndexUnsafe 0 arr
+                                 lstElem = Array.getIndexUnsafe (len - 1) arr
+                                 w16 = (remElem, fstElem)
```

Comment:
This will fail if the array is empty.
1. We need a test case that catches that.
2. Fix it.